internal/daemon/api_files.go

```diff
@@ -0,0 +1,576 @@
+
+type fileResult struct {
+	Path  string       `json:"path"`
+	Error *errorResult `json:"error,omitempty"`
+}
+
+// Reading files
+
+// Custom Response implementation to serve the multipart.
+type readFilesResponse struct {
+	paths []string
+}
+
+func (r readFilesResponse) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Write HTTP status and headers. HTTP status is always OK because we don't
+	// know any better until we've read all the files.
+	mw := multipart.NewWriter(w)
+	header := w.Header()
+	header.Set("Content-Type", mw.FormDataContentType())
+	w.WriteHeader(http.StatusOK)
+
+	// Read each file's contents to multipart response.
+	result := make([]fileResult, len(r.paths))
+	for i, path := range r.paths {
+		err := readFile(path, mw)
+		result[i] = fileResult{
+			Path:  path,
+			Error: fileErrorToResult(err),
+		}
+	}
+
+	// At the end, write response metadata in JSON format.
+	mh := textproto.MIMEHeader{}
+	mh.Set("Content-Type", "application/json")
+	mh.Set("Content-Disposition", `form-data; name="response"`)
+	part, err := mw.CreatePart(mh)
+	if err != nil {
+		// Can't write metadata -- writing the error message on a new line is
+		// about the best we can do.
+		fmt.Fprint(w, "\n", err)
+		return
+	}
+	encoder := json.NewEncoder(part)
+	metadata := respJSON{
+		Type:       ResponseTypeSync,
+		Status:     http.StatusOK,
+		StatusText: http.StatusText(http.StatusOK),
+		Result:     result,
+	}
+	err = encoder.Encode(metadata)
+	if err != nil {
+		fmt.Fprint(w, "\n", err)
+		return
+	}
+
+	// Write the multipart trailer (last boundary marker).
+	err = mw.Close()
+	if err != nil {
+		fmt.Fprint(w, "\n", err)
+		return
+	}
+}
+
+func nonAbsolutePathError(path string) error {
+	return fmt.Errorf("paths must be absolute, got %q", path)
+}
+
+func readFile(path string, mw *multipart.Writer) error {
+	if !pathpkg.IsAbs(path) {
+		return nonAbsolutePathError(path)
+	}
+	if osutil.IsDir(path) {
+		return fmt.Errorf("cannot read a directory: %q", path)
+	}
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	fw, err := mw.CreateFormFile("files", path)
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(fw, f)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func fileErrorToResult(err error) *errorResult {
+	if err == nil {
+		return nil
+	}
+	var kind errorKind
+	switch {
+	case errors.Is(err, os.ErrPermission):
+		kind = errorKindPermissionDenied
+	case errors.Is(err, os.ErrNotExist):
+		kind = errorKindNotFound
+	default:
+		kind = errorKindGenericFileError
+	}
+	return &errorResult{
+		Kind:    kind,
+		Message: err.Error(),
+	}
+}
+
+// Listing files
+
+func listErrorResponse(err error) Response {
+	status := http.StatusBadRequest
+	kind := errorKindGenericFileError
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		status = http.StatusNotFound
+		kind = errorKindNotFound
+	case errors.Is(err, os.ErrPermission):
+		status = http.StatusForbidden
+		kind = errorKindPermissionDenied
+	}
```

Comment:
The function above is entirely contained inside this function. We can just `return &resp{..., fileErrorToResult(...)}` I think.

Reply:
Refactored/simplified - thanks. (We still need the status here.)